src/style-spec/function/expression.js

```diff
@@ -80,39 +73,22 @@ class BaseExpression {
 
 class LiteralExpression extends BaseExpression {
     value: Value;
-    constructor(key: *, type: PrimitiveType | ArrayType, value: Value) {
+
+    constructor(key: *, type: Type, value: Value) {
         super(key, type);
-        (this: any).value = value;
+        this.value = value;
     }
 
-    static inferArrayType(value: Array<Value>) {
-        let itemType;
-        // infer the array's item type
-        for (const item of value) {
-            const t = primitiveTypes[typeof item];
-            if (t && !itemType) {
-                itemType = t;
-            } else if (t && itemType === t) {
-                continue;
-            } else {
-                itemType = ValueType;
-                break;
-            }
-        }
+    static parse(args: Array<mixed>, context: ParsingContext) {
+        if (args.length !== 1)
+            throw new ParsingError(context.key, `'literal' expression requires exactly one argument, but found ${args.length} instead.`);
 
-        return array(itemType || ValueType, value.length);
-    }
+        if (!isValue(args[0]))
```

Comment:
As of now, there's actually no way to express a color literal -- `rgba` and `color` are functions -- but this would allow `['literal', new Color(...)]`.